R/missingLevelsTrain.R

```diff
@@ -0,0 +1,70 @@
+#' @title missingLevelsTrain
+#' @description
+#' Accounts for missing factor levels only present in test data
+#' but not in train data by setting levels to NA
+#' @param fit \cr
+#'   fitted model on training data
+#' @param test.data [`data.frame`] \cr
+#'   data to make predictions for
+#' @return `data.frame` with matching factor levels to fitted model
+#' @keywords internal
+#' @export
+missingLevelsTrain = function(fit, test.data) {
+
+  # https://stackoverflow.com/a/39495480/4185785
+
+  # drop empty factor levels in test data
+  test.data = as.data.frame(droplevels(test.data))
+
+  # 'fit' object structure of 'lm' and 'glmmPQL' is different so we need to
+  # account for it
+  if (any(class(fit) == "glmmPQL")) {
+    # Obtain factor predictors in the model and their levels
+    factors = stri_replace_all_regex(names(unlist(fit$learner.model$contrasts)),
+                                     "[-^0-9]|as.factor|\\(|\\)", "")
+    # do nothing if no factors are present
+    if (length(factors) == 0) {
+      return(test.data)
+    }
+
+    factor.levels = unlist(lapply(fit$contrasts, function(x)
+      outer(rownames(x), colnames(x), paste, sep = ":")))
+    factor.levels = stri_split(factor.levels, ":", simplify = TRUE)[, 1]
+
+    model.factors = as.data.frame(cbind(factors, factor.levels))
+  } else {
+    # Obtain factor predictors in the model and their levels
+    factors = stri_replace_all_regex(names(unlist(fit$learner.model$xlevels)),
+                                     "[-^0-9]|as.factor|\\(|\\)", "")
+
+    if (length(factors) == 0) {
+      return(test.data)
+    }
+    factor.levels = unname(unlist(fit$learner.model$xlevels))
```

Comment:
Where do the xlevels come from?

Reply:
Its a sub-object of a fitted lm* model and contains the factor levels of the trained model